File: leetcode/algorithms/49_group_anagrams/main.js
/*
 * Complexities:
 *   N - size of strs
 *   K - max size of strings in strs
 *   - Time Complexity: O(N * K * logᴷ)
 *   - Space Complexity: O(N * K)
 */
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const strMap = new Map();

  for (let i = 0; i < strs.length; i++) {
    const sortedStr = strs[i].split("").sort().join("");

    if (!strMap.has(sortedStr)) {
      strMap.set(sortedStr, new Array());
    }

    strMap.get(sortedStr).push(strs[i]);
  }

  return Array.from(strMap.values());
};


// Solution
/**
 * Solution 1
 *
 * Sort & Group
 *
 * Complexities:
 *   N - size of strs
 *   K - max size of strings in strs
 *   - Time Complexity: O(N * K * logᴷ)
 *   - Space Complexity: O(N * K)
 */
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var solution1 = function (strs) {
  const groups = {};

  for (const str of strs) {
    const key = str.split("").sort().join("");

    if (!groups[key]) {
      groups[key] = [];
    }

    groups[key].push(str);
  }

  return Object.values(groups);
};

/**
 * Solution 2
 *
 * Count & Group
 *
 * Complexities:
 *   N - size of strs
 *   K - max size of strings in strs
 *   - Time Complexity: O(N * K)
 *   - Space Complexity: O(N * K)
 */
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var solution2 = function (strs) {
  const groups = {};
  const charCodeOfA = "a".charCodeAt(0);

  for (const str of strs) {
    const count = new Array(26).fill(0);

    for (let i = 0; i < str.length; i++) {
      const index = str.charCodeAt(i) - charCodeOfA;
      count[index]++;
    }

    const key = count.join("#");

    if (!groups[key]) {
      groups[key] = [];
    }
    groups[key].push(str);
  }

  return Object.values(groups);
};

File: leetcode/algorithms/49_group_anagrams/main.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class GroupAnagrams {
    /**
     * Complexities:
     * N - size of strs
     * K - max size of strings in strs
     * - Time Complexity: O(N * K * logᴷ)
     * - Space Complexity: O(N * K)
     */
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> strMap = new HashMap<>();

        for (String str : strs) {
            char[] strArray = str.toCharArray();
            Arrays.sort(strArray);
            String sortedStr = new String(strArray);

            if (!strMap.containsKey(sortedStr)) {
                strMap.put(sortedStr, new ArrayList<>());
            }

            strMap.get(sortedStr).add(str);
        }

        return new ArrayList<>(strMap.values());
    }


    // Solution
    /**
     * Solution 1
     * 
     * Sorted String as Key
     * 
     * Complexities:
     * N - size of strs
     * K - max size of strings in strs
     * - Time Complexity: O(N * K * logᴷ)
     * - Space Complexity: O(N * K)
     */
    public List<List<String>> solution1(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            char[] charArray = s.toCharArray();

            Arrays.sort(charArray);

            String key = new String(charArray);

            map.computeIfAbsent(key, k -> new ArrayList<>());
            map.get(key).add(s);
        }

        return new ArrayList<>(map.values());
    }

    /**
     * Solution 2
     * 
     * Character Count as Key
     * 
     * Complexities:
     * N - size of strs
     * K - max size of strings in strs
     * - Time Complexity: O(N * K)
     * - Space Complexity: O(N * K)
     */
    public List<List<String>> solution2(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            char[] counts = new char[26];

            for (char c : s.toCharArray()) {
                counts[c - 'a']++;
            }

            String key = new String(counts);

            map.computeIfAbsent(key, k -> new ArrayList<>());
            map.get(key).add(s);
        }

        return new ArrayList<>(map.values());
    }
}
